### ISSUE TITLE
Removal of locks from SNITcpHandle.cs breaks EF tests

### ISSUE THREAD

**Issue by saurabh500** [dotnet/corefx#27587]
Removal of locks in the PR https://github.com/dotnet/corefx/pull/27187 breaks EF tests with the following stack trace. To trace this failure to the lock removal I undid the changes in the PR and then just removed the locks without any allocation optimizations.
I could repro this by running EF tests on Windows with Managed SNI and on Linux as well
This could potentially be a timing issue and could repro on Linux in different scenarios.

```
[xUnit.net 00:01:45.7938557]     Microsoft.EntityFrameworkCore.Query.QueryBugsTest.Repro5456_multiple_include_group_join_is_per_query_context_async [FAIL]
[xUnit.net 00:01:45.7967113]       System.InvalidOperationException : Invalid attempt to call ReadAsync when reader is closed.
[xUnit.net 00:01:45.7990081]       Stack Trace:
[xUnit.net 00:01:45.8012431]            at System.Data.Common.ADP.ExceptionWithStackTrace(Exception e)
[xUnit.net 00:01:45.8017075]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8068374]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8070711]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8073990]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8075715]         D:\aspnet\EntityFrameworkCore\src\EFCore.Relational\Query\Internal\AsyncQueryingEnumerable.cs(138,0): at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<BufferlessMoveNext>d__12.MoveNext()
[xUnit.net 00:01:45.8078974]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8080208]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8082113]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8083356]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8084510]         D:\aspnet\EntityFrameworkCore\src\EFCore.SqlServer\Storage\Internal\SqlServerExecutionStrategy.cs(66,0): at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.<ExecuteAsync>d__7`2.MoveNext()
[xUnit.net 00:01:45.8086315]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8087293]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8088197]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8089066]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8090061]         D:\aspnet\EntityFrameworkCore\src\EFCore.Relational\Query\Internal\AsyncQueryingEnumerable.cs(84,0): at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<MoveNext>d__11.MoveNext()
[xUnit.net 00:01:45.8091798]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8092741]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8136308]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8138275]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8139079]         D:\aspnet\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs(1252,0): at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext()
[xUnit.net 00:01:45.8139669]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8140505]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8141557]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8142535]            at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
[xUnit.net 00:01:45.8143464]            at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.<MoveNextCore>d__7.MoveNext()
[xUnit.net 00:01:45.8144448]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8145468]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8146503]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8149178]            at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
[xUnit.net 00:01:45.8150135]            at System.Linq.AsyncEnumerable.AsyncIterator`1.<MoveNext>d__10.MoveNext()
[xUnit.net 00:01:45.8150609]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8151621]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8153297]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8154215]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8155213]         D:\aspnet\EntityFrameworkCore\src\EFCore\Query\Internal\AsyncLinqOperatorProvider.cs(157,0): at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.<MoveNext>d__5.MoveNext()
[xUnit.net 00:01:45.8155743]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8156111]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8156864]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8157309]            at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
[xUnit.net 00:01:45.8158184]            at System.Linq.AsyncEnumerable.<Aggregate_>d__6`3.MoveNext()
[xUnit.net 00:01:45.8159034]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8159504]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8159906]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8160622]            at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
[xUnit.net 00:01:45.8161423]         D:\aspnet\EntityFrameworkCore\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs(1559,0): at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.<<Repro5456_multiple_include_group_join_is_per_query_context_async>b__83_0>d.MoveNext()
[xUnit.net 00:01:45.8163444]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8164334]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8165463]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8166412]         D:\aspnet\EntityFrameworkCore\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs(1552,0): at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.<Repro5456_multiple_include_group_join_is_per_query_context_async>d__83.MoveNext()
[xUnit.net 00:01:45.8167330]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8168143]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8168972]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8169579]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8170187]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8170617]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[xUnit.net 00:01:45.8171179]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:01:45.8171815]            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[xUnit.net 00:01:45.8172482]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Failed   Microsoft.EntityFrameworkCore.Query.QueryBugsTest.Repro5456_multiple_include_group_join_is_per_query_context_async
Error Message:
 System.InvalidOperationException : Invalid attempt to call ReadAsync when reader is closed.
Stack Trace:
   at System.Data.Common.ADP.ExceptionWithStackTrace(Exception e)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<BufferlessMoveNext>d__12.MoveNext() in D:\aspnet\EntityFrameworkCore\src\EFCore.Relational\Query\Internal\AsyncQueryingEnumerable.cs:line 138
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.<ExecuteAsync>d__7`2.MoveNext() in D:\aspnet\EntityFrameworkCore\src\EFCore.SqlServer\Storage\Internal\SqlServerExecutionStrategy.cs:line 66
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.<MoveNext>d__11.MoveNext() in D:\aspnet\EntityFrameworkCore\src\EFCore.Relational\Query\Internal\AsyncQueryingEnumerable.cs:line 84
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext() in D:\aspnet\EntityFrameworkCore\src\EFCore\Query\EntityQueryModelVisitor.cs:line 1252
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.<MoveNextCore>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Linq.AsyncEnumerable.AsyncIterator`1.<MoveNext>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.<MoveNext>d__5.MoveNext() in D:\aspnet\EntityFrameworkCore\src\EFCore\Query\Internal\AsyncLinqOperatorProvider.cs:line 157
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Linq.AsyncEnumerable.<Aggregate_>d__6`3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.<<Repro5456_multiple_include_group_join_is_per_query_context_async>b__83_0>d.MoveNext() in D:\aspnet\EntityFrameworkCore\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs:line 1559
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.EntityFrameworkCore.Query.QueryBugsTest.<Repro5456_multiple_include_group_join_is_per_query_context_async>d__83.MoveNext() in D:\aspnet\EntityFrameworkCore\test\EFCore.SqlServer.FunctionalTests\Query\QueryBugsTest.cs:line 1552
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by danmosemsft** [comment]
@benaadams didn't you check in a change to eliminate all this async stack clutter?

**Comment by saurabh500** [comment]
Closing this issue. Fixed with https://github.com/dotnet/corefx/pull/27591

**Comment by stephentoub** [comment]
> didn't you check in a change to eliminate all this async stack clutter?

@saurabh500, what coreclr was being used when you got that stack trace?

**Comment by benaadams** [comment]
>  didn't you check in a change to eliminate all this async stack clutter?

coreclr preview1 removes

```
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
```
coreclr preview2 early additionally removes
```
--- End of stack trace from previous location where exception was thrown ---
```
coreclr preview2 later additionally converts to its correct signiture
```
at System.Linq.AsyncEnumerable.<Aggregate_>d__6`3.MoveNext()
```

Was this error on netfx?

**Comment by saurabh500** [comment]
```
.NET Command Line Tools (2.2.0-preview1-007866)

Product Information:
 Version:            2.2.0-preview1-007866
 Commit SHA-1 hash:  a99154ed46

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /home/saurabh/.dotnet/sdk/2.2.0-preview1-007866/

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview2-26131-06
  Build    : b13a0d5c331f374afd35ded57b9a4b4ab128864c
```

I have been using the EF builds to run these tests. So the CoreCLR is dependant on the version of SDK and CLI. The above is the output of `dotnet --info`

**Comment by benaadams** [comment]
`2.1.0-preview2-26131-06` had a GC hole https://github.com/dotnet/coreclr/pull/16231; was fixed in `2.1.0-preview2-26207-11`
